docs/core/dependency-loading/understanding-assemblyloadcontext.md

```diff
@@ -0,0 +1,81 @@
+[Unmanaged (native) library loading algorithm](loading-unmanaged.md) refer to all the available events and virtual functions.  The articles show each event and functions relative position in the loading algorithms. This article won't reproduce that information.
+
+This section will cover the general principles for the relevant events and functions.
+
+- **Be repeatable**. It's required that a query for a specific dependency will always result in the same response. The same loaded dependency instance must be returned. This requirement is fundamental  for cache consistency.
+  - The exception is that the initial `null` response could be followed by a later non-`null` response.
+  - For managed assemblies in particular, we're creating a <xref:System.Reflection.AssemblyName.Name?displayProperty=nameWithType> to <xref:System.Reflection.Assembly> cache. There can be exactly one <xref:System.Reflection.AssemblyName.Name?displayProperty=nameWithType> in the cache.
+- **Typically don't throw**.  It's expected that these functions return `null` rather than throw when unable to find the requested dependency. Throwing should be restricted to unexpected errors like a corrupt assembly or out of memory.
+- **Avoid recursion**. Beware that these functions and handlers are implementing the loading rules for locating dependencies. Your implementation shouldn't call APIs that trigger recursion. Your code should typically be calling **AssemblyLoadContext** load functions that require a specific path or memory reference argument.
+- **Load into the correct AssemblyLoadContext**. The choice of where to load dependencies is application-specific.  The choice is implemented by these events and functions. When your code calls **AssemblyLoadContext** load-by-path functions call them on the instance where you want the code loaded. Sometime returning `null` and letting the <xref:System.Runtime.Loader.AssemblyLoadContext.Default?displayProperty=nameWithType> handle the load may be the simplest option.
+
+## How are dynamic dependencies isolated?
+
+One of the primary purposes of exposing the <xref:System.Runtime.Loader.AssemblyLoadContext> is to allow for isolation of dependencies.
+
+The general static dependency model requires a one to one mapping of name to dependency instance.
+
+For dynamic dependencies, there's no guarantee that another static or dynamic dependency hasn't already loaded a dependency of the same name.
+
+Each <xref:System.Runtime.Loader.AssemblyLoadContext> instance represents a unique scope for mapping names.
+
+There's no binary isolation between these dependencies. They're only isolated by not finding each other by name.
+
+## How are dependencies shared?
+
+Dependencies can easily be shared between <xref:System.Runtime.Loader.AssemblyLoadContext> instances. The general model is for one <xref:System.Runtime.Loader.AssemblyLoadContext> to load a dependency.  The other share the dependency by adding the dependency instance to their cache.
```

Comment:
This is misleading - sharing dependencies is done by "loading" them through the same ALC. The "other" ALCs never add them to their cache. Maybe a better way to describe this would be that you share them by loading them just once (which is the outcome of loading them through the same ALC).

Reply:
> This is misleading

I'll reread it.

>  The "other" ALCs never add them to their cache. 

For shared assemblies, an assembly reference is added to both ALCs' caches. The cache is just a simple name to assembly dictionary. The assembly has a data member which knows which ALC loaded it.

Reply:
Didn't know that - thanks for clarification!